pkg/restapi/webhook/webhook.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/aries-framework-go/pkg/common/log"
+)
+
+type webhookRequest struct {
+	Name        string
+	Destination string
+}
+
+var logger = log.New("aries-framework/agentd")
+
+// UnexpectedExitErrMsg is the string to be displayed when the webhook server exits unexpectedly.
+const UnexpectedExitErrMsg = "webhook server exited unexpectedly"
+
+// StartWebhookServer will start up a new webhook server at the given host URL.
+// The handlerPattern is appended to the host URL.
+// e.g. localhost:8080/webhook
+func StartWebhookServer(host, handlerPattern string) error {
+	dispatcher := &dispatcher{
+		client:         &http.Client{},
+		subscriberURLs: make(map[string]string),
+		mu:             &sync.Mutex{},
+	}
+
+	go dispatcher.start()
+	logger.Infof("Subscribe to webhook at %s%s \n", host, handlerPattern)
+
+	return listenForNewSubscribers(host, handlerPattern, dispatcher)
+}
+
+func listenForNewSubscribers(host, handlerPattern string, dispatcher *dispatcher) error {
+	srv := &http.Server{Addr: host, Handler: http.DefaultServeMux}
+
+	http.HandleFunc(handlerPattern, func(resp http.ResponseWriter, req *http.Request) {
+		addNewWebhookSubscriber(req, resp, dispatcher)
+	})
+
+	err := srv.ListenAndServe()
+	if err != nil {
+		return fmt.Errorf("webhook server failed to start: %s", err)
+	}
+	return nil
+}
+
+func addNewWebhookSubscriber(req *http.Request, resp http.ResponseWriter, dispatcher *dispatcher) {
+	dec := json.NewDecoder(req.Body)
+	var request webhookRequest
+	err := dec.Decode(&request)
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		logger.Errorf("Error adding new subscriber: %s", err)
+		return
+	}
+	dispatcher.addSubscriber(request.Name, request.Destination)
+}
+
+type dispatcher struct {
+	client         *http.Client
+	subscriberURLs map[string]string
+	mu             *sync.Mutex
+}
+
+func (dispatcher *dispatcher) start() {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for range ticker.C {
+		dispatcher.dispatch()
+	}
+}
+func (dispatcher *dispatcher) addSubscriber(name, subscriberURL string) {
```

Comment:
to improve readability -->use  func (d *dispatcher

Reply:
Done!